### DISCUSSION TITLE
Release from tagged pushes

### DISCUSSION THREAD

**Discussion by PloxKevin** [community/community#25974] [Actions]
<p>Hey everyone, I recently got into CI and saw the way it allows for Artefacts to be used. What I would like to do is:</p>
<ol>
<li>If I push a commit, that it builds the project (C#). If it is succesful, I can use the tag. This step I’ve figured out.</li>
<li>If I push a tagged commit, it would be nice that it also posts a release version that people could download. I use GitHub desktop, and you have those two boxes. One for summary and one for description. If those would be usable in creating a Title and description for the release, it would be great. (Working on this currently)</li>
<li>Is it possible for the release number to increment. So v1.1, v1.2 etc?</li>
</ol>
<p>I am currently on 30ish failed updated .yml files.</p>
<p>The project I am working on is: <a href="https://github.com/Laemers/BattleNavHelper" rel="nofollow noopener">https://github.com/Laemers/BattleNavHelper</a><br>
My <code>.yml</code> is as follows:</p>
<pre><code class="lang-auto">name: .NET Core

on:
  push:
    branches: [ master ]
  pull_request:
    branches: [ master ]

jobs:
  build:

    runs-on: windows-latest

    steps:
    - uses: actions/checkout@v2
    - name: Setup Nuget.exe
      uses: warrenbuckley/Setup-Nuget@v1
    - name: Restore Package
      run: nuget restore BattleNavHelper.sln 
    - name: Setup MSBuild.exe
      uses: microsoft/Setup-MSBuild@v1
    - name: Build with MSBuild
      run: msbuild BattleNavHelper.sln

  release:
    name: Create Release
    runs-on: windows-latest
    steps:
      - name: Checkout code
        uses: actions/checkout@v2
      - name: Create Release
        id: create_release
        uses: actions/create-release@v1
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }} # This token is provided by Actions, you do not need to create your own token
        with:
          tag_name: ${{ github.ref }}
          release_name: Release ${{ github.ref }}
          body: |
            Changes in this Release
            - First Change
            - Second Change
          draft: false
          prerelease: false
</code></pre>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/ploxkevin">@PloxKevin</a>,</p>
<p>By default, the <a href="https://help.github.com/en/actions/reference/events-that-trigger-workflows#push-event-push" rel="nofollow noopener"><strong>push event</strong></a> is available to pushes to branches and pushes to tags.</p>
<pre><code class="lang-yaml">on:
  push:  # no any branches filter or tags filter set on push, all pushes can triggered the workflow
</code></pre>
<p>However, if you only define branches filter (<strong>branches</strong> or <strong>branches-ignore</strong>), the workflow won’t be triggered when you push to tags. similarly, if your only define tags filter (<strong>tags</strong> or <strong>tags-ignore</strong>), the workflow won’t be triggered when you push to branches.</p>
<p>If you want the workflow will be triggered when you push to some specified branches or tags, you need to add both branches filter and tags filter on push event.</p>
<pre><code class="lang-yaml">on:
  push:
    branches:
      - 'branches_filter_pattern_1'
      - 'branches_filter_pattern_2'
      . . .
      - 'branches_filter_pattern_N'
    tags:
      - 'tags_filter_pattern_1'
      - 'tags_filter_pattern_2'
      . . .
      - 'tags_filter_pattern_N'
</code></pre>
<p>About generating auto-incrementing build/release number, I found the following actions from the <a href="https://github.com/marketplace?type=actions&amp;query=build+number" rel="nofollow noopener"><strong>GitHub Marketplace</strong></a>, you can use one of them in your workflow:</p>
<ul>
<li><a href="https://github.com/marketplace/actions/build-number" rel="nofollow noopener"><strong>Build Number</strong></a></li>
<li><a href="https://github.com/marketplace/actions/build-number-generator" rel="nofollow noopener"><strong>Build Number Generator</strong></a></li>
<li><a href="https://github.com/marketplace/actions/gh-action-buildnum" rel="nofollow noopener"><strong>gh-action-buildnum</strong></a></li>
</ul>
<p>According to your requests, you can reference the demo below to configure your workflow.</p>
<pre><code class="lang-yaml">on:
  push:
    branches:
      &lt;some branches_filter_patterns&gt;
    tags:
      &lt;some tags_filter_patterns&gt;
  pull_request:
    branches:
      &lt;some branches_filter_patterns&gt;
    tags:
      &lt;some tags_filter_patterns&gt;

jobs:
  build:
    name: Build project
    . . .
    steps:
    ---- some steps for build ----

  get_tag:
    name: Get release tag
    needs: build
    runs-on: ubuntu-latest
    outputs:
      release_tag: ${{ steps.output_tag.outputs.tag }}
    steps:
      - name: Get existing tag from github.ref
        if: ${{ contains(github.ref, 'refs/tags/') }}
        run: echo "::set-env name=tag::${GITHUB_REF#refs/tags/}"

      - name: Generate number
        id: generate_number
        if: ${{ contains(github.ref, 'refs/heads/') }}
        uses: travisj/build-number-action@v1

      - name: Generate new tag
        if: ${{ contains(github.ref, 'refs/heads/') }}
        run: echo "::set-env name=tag::v1.${{ steps.generate_number.outputs.build-number }}"

      - name: Set output
        id: output_tag
        run: echo "::set-output name=tag::${{ env.tag }}"

  release:
    name: Release project
    needs: get_tag
    . . .
    steps:
      . . .
      - name: Create Release
        id: create_release
        uses: actions/create-release@v1
        with:
          tag_name: ${{ needs.get_tag.outputs.release_tag }}
          release_name: Release ${{ needs.get_tag.outputs.release_tag }}
          body: |
           . . .
      . . .
</code></pre>

**Comment by PloxKevin** [comment]
<p>Hey <a class="mention" href="/brightran">@brightran</a>.</p>
<p>Thank you so much for helping me on the last part. I have advanced a bit now and I currently the following. But first, let me explain the situation:</p>
<ol>
<li>I use GitHub desktop and I use right click on the commits to add a tag to a certain commit.</li>
</ol>
<p>The <code>build</code> step and the <code>get_tag</code> step work perfectly fine. They generate a pass. Currently, in the code it allows for the artifact to be uploaded. So I am pretty close to my final personal goal.</p>
<p>So when I push using a tag, say for instance <code>v1.3.2</code>, it still gives me the error : <code>This is not a tagged push.</code>.<br>
I figured that I need to fix my code in such a way that it needs to work with manual tags first, and when that works, I could look into auto-generating tags. From there on I can use <code>if</code> statements to check when it needs to be released or not. My code is as follows.</p>
<p>Maybe it’s a problem as I use windows-latest on the build part and ubuntu for the other parts?</p>
<pre><code class="lang-auto">name: .NET Core

on:
  push:
    branches: [ master ]
  pull_request:
    branches: [ master ]

jobs:
  build:
    runs-on: windows-latest
    
    steps:
    - uses: actions/checkout@v2
    - name: Setup Nuget.exe
      uses: warrenbuckley/Setup-Nuget@v1
    - name: Restore Package
      run: nuget restore BattleNavHelper.sln
    - name: Setup MSBuild.exe
      uses: microsoft/Setup-MSBuild@v1
    - name: Build with MSBuild
      run: msbuild BattleNavHelper.sln
    - run: mkdir staging &amp;&amp; cp BattleNavHelper/bin/Debug/*.exe staging
    - uses: actions/upload-artifact@v1
      with:
        name: Package
        path: staging
        
  get_tag:
    name: Get release tag
    needs: build
    runs-on: ubuntu-latest
    outputs:
      release_tag: ${{ steps.output_tag.outputs.tag }}
    steps:
      - name: Get existing tag from github.ref
        if: ${{ contains(github.ref, 'refs/tags/') }}
        run: echo "::set-env name=tag::${GITHUB_REF#refs/tags/}"

      - name: Generate number
        id: generate_number
        if: ${{ contains(github.ref, 'refs/heads/') }}
        uses: travisj/build-number-action@v1

      - name: Generate new tag
        if: ${{ contains(github.ref, 'refs/heads/') }}
        run: echo "::set-env name=tag::v1.${{ steps.generate_number.outputs.build-number }}"

      - name: Set output
        id: output_tag
        run: echo "::set-output name=tag::${{ env.tag }}"
        
  release:
    runs-on: ubuntu-latest
    needs: [build, get_tag]
    steps:
      - uses: actions/checkout@master
      - name: Create GitHub release
        uses: Roang-zero1/github-create-release-action@master
        with:
          version_regex: '^v[[:digit:]]+\.[[:digit:]]+\.[[:digit:]]+'
          update_existing: 'true'
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
</code></pre>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/ploxkevin">@PloxKevin</a>,</p>
<blockquote>
<p>So when I push using a tag, say for instance  <strong>v1.3.2</strong>, it still gives me the error :  <strong>This is not a tagged push.</strong>.</p>
</blockquote>
<p>Which step do you get this error in? Manually create tag on <strong>GitHub Desktop</strong>, or run some step in the workflow?</p>
<blockquote>
<p><strong>Maybe it’s a problem as I use windows-latest on the build part and ubuntu for the other parts?</strong></p>
</blockquote>
<p>As far as your case is concerned, I don’t think these will have any affect.</p>
<p>In addition, GitHub has provided an official action for users to create release (<a href="https://github.com/actions/create-release" rel="nofollow noopener"><strong>actions/create-release</strong></a>), you can directly use this action to create a release with the tag that you get from the “<strong>get_tag</strong>” job.</p>

**Comment by PloxKevin** [comment]
<p><a class="mention" href="/brightran">@brightran</a></p>
<p>The error is given in the workflow. Adding a tag on Github desktop works, and allows me to push this change. Since I am not filtering for branches or tags, the workflow should trigger on each commit.</p>
<p>So what I did was, create commit, add a tag, and then push it using Github Desktop.</p>

**Comment by BrightRan** [comment, marked as answer]
<p><a class="mention" href="/ploxkevin">@PloxKevin</a>,</p>
<blockquote>
<p>Since I am not filtering for branches or tags, the workflow should trigger on each commit.</p>
</blockquote>
<p>But I see you have set branches filter on push event in your workflow.</p>
<pre><code class="lang-yaml">on:
  push:
    branches: [ master ]
  . . .
</code></pre>
<p>This will trigger the workflow when you push commits to master branch. Pushes to tags will not trigger the workflow. As I mentioned abouve,</p>
<blockquote>
<p>if you only define branches filter ( <strong>branches</strong>  or  <strong>branches-ignore</strong> ), the workflow won’t be triggered when you push to tags. similarly, if your only define tags filter ( <strong>tags</strong>  or  <strong>tags-ignore</strong> ), the workflow won’t be triggered when you push to branches.</p>
</blockquote>
<p>When you push commits to master branch, the git ref (<strong>github.ref</strong>) is ‘<strong>refs/heads/master</strong>’, it is a branch ref, not tag ref. You can’t directly release a branch ref.</p>
<p>You can try like as below:</p>
<pre><code class="lang-yaml">on:
  push:
  . . .
</code></pre>
<p>OR</p>
<pre><code class="lang-yaml">on:
  push:
    branches:
      - master
    tags:
      - '**'
  . . .
</code></pre>
<p>In addition, you have get the tag name from the “<strong>get_tag</strong>” job, you can directly use this tag name to release.</p>
